pkg/client/client_test.go

```diff
@@ -1,56 +1,660 @@
 package client
 
 import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
 	"net/http"
+	"path"
+	"strings"
 	"testing"
 
-	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/runtime/serializer"
-	"k8s.io/client-go/rest"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest/fake"
 )
 
-type userAgentRoundTripper struct {
-	expectedUserAgent string
-	t                 *testing.T
+var accessor = meta.NewAccessor()
+var testSchema = runtime.NewScheme()
+var negotiatedSerializer = serializer.NewCodecFactory(testSchema)
+
+func TestMain(m *testing.M) {
+	// Setup Schemas for serializer before we run our test
+	metav1.AddMetaToScheme(testSchema)
+	v1.SchemeBuilder.AddToScheme(testSchema)
+	m.Run()
+}
+
+// Object is generic object to wrap runtime and metav1.
+type Object interface {
+	metav1.Object
+	runtime.Object
+}
+
+type testInfo struct {
+	name    string
+	fields  clientFields
+	args    testArgs
+	wantErr bool
+}
+type clientFields struct {
+	Namespaced bool
+	GVR        schema.GroupVersionResource
+	Kind       string
+}
+type testArgs struct {
+	obj       Object
+	namespace string
+	desired   runtime.Object
+	result    runtime.Object
+}
+
+func TestClient_Get(t *testing.T) {
```

Comment:
In addition to the client operations functions, I would recommend that you make some tests for `WithAgent` and `WithImpersonation`. Particularly since we need them to work well together (i.e. not overwrite the config).

Reply:
Thanks for adding these - they mostly look good. I did have one question on one assert here: https://github.com/rancher/lasso/blob/9641c9272c6ad585e58f3ba64d4df83056485805/pkg/client/client_test.go#L659 - The test previously asserts that `newClient.Config.UserAgent` is the new user agent. I would expect that the original config is not equal to this value, so `newClient.Config.UserAgent == agentName` and `originalConfig.UserAgent != agentName`, `newClient.Config != originalConfig`. Would you mind clarifying what I am missing?

Reply:
You are correct the line above this set's the user-agents on the newClient. `newClient.Config.UserAgent = originalConfig.UserAgent`

Reply:
I see, so you are reverting the user agent to check for equality. That makes sense, resolving.